### PR TITLE
feat: adds support for bulk operations

### PR DESCRIPTION
**Pull request by farafonov-alexey**
This PR adds support for the endpoints:

1. PUT /{userId}/orgs/{orgId}/roles - input json: `List<OrganizationRole>`
Grant organization roles for user

2. DELETE /{userId}/orgs/{orgId}/roles - input  json `List<OrganizationRole>`
Revoke organization roles for user

4. PUT /orgs/{orgId}/roles - input json `List<OrganizationRole>`
Create organization roles

6. DELETE /orgs/{orgId}/roles - input json `List<OrganizationRole>`
Delete organization roles

8. GET orgs/{orgId}/members/with-roles
Query params the same as "orgs/:orgId/members"
Returns UserRepresentationWithRoles -` { user: UserRepresentation, organizationRoles: List<OrganizationRole> }`

### CODE REVIEW

**Comment by farafonov-alexey** [comment]
The last endpoint here not quite suitable for this PR. But we need this feature too =)

File: src/main/java/io/phasetwo/service/resource/MembersResource.java [hunk 1]
```diff
@@ -40,6 +44,29 @@ public Stream<UserRepresentation> getMembers(
         .searchForMembersStream(searchQuery, firstResult, maxResults)
         .map(m -> toRepresentation(session, realm, m));
   }
+  @GET
+  @Path("/with-roles")
```
Comment (xgp) [inline src/main/java/io/phasetwo/service/resource/MembersResource.java:48]: Won't take this one. Rationale is that it's only one additional call, and I don't want to pollute the API with convenience methods and objects.

File: src/main/java/io/phasetwo/service/resource/RolesResource.java [hunk 2]
```diff
@@ -41,12 +44,52 @@ public Stream<OrganizationRole> getRoles() {
...
+            session.getContext().getUri().getAbsolutePathBuilder().path(or.getName()).build())
+        .build();
+  }
+
+  @PUT
+  @Consumes(MediaType.APPLICATION_JSON)
+  public Response createRoles(List<OrganizationRole> representation) {
+    canManage();
+
+    representation.forEach(this::createOrganizationRole);
```
Comment (xgp) [inline src/main/java/io/phasetwo/service/resource/RolesResource.java:61]: Need a test that validates when some of the creates fail:
1. everything fails
2. no events get emitted
3. a proper error gets sent back

File: src/main/java/io/phasetwo/service/resource/RolesResource.java [hunk 3]
```diff
@@ -41,12 +44,52 @@ public Stream<OrganizationRole> getRoles() {
...
+
```
Comment (xgp) [inline src/main/java/io/phasetwo/service/resource/RolesResource.java:62]: Should have a good error showing which role create failed.

File: src/main/java/io/phasetwo/service/resource/RolesResource.java [hunk 4]
```diff
@@ -41,12 +44,52 @@ public Stream<OrganizationRole> getRoles() {
...
+      String name = r.getName();
+
+      if (Arrays.asList(OrganizationAdminAuth.DEFAULT_ORG_ROLES).contains(name)) {
+        throw new BadRequestException(
+                String.format("Default organization role %s cannot be deleted.", name));
+      }
+
+      organization.removeRole(name);
+
+      adminEvent
```
Comment (xgp) [inline src/main/java/io/phasetwo/service/resource/RolesResource.java:82]: Need a test that validates we can emit multiple events per request.

File: src/main/java/io/phasetwo/service/resource/RolesResource.java [hunk 5]
```diff
@@ -57,14 +100,21 @@ public Response createRole(OrganizationRole representation) {
 
     OrganizationRole or = convertOrganizationRole(r);
     adminEvent
-        .resource(ORGANIZATION_ROLE.name())
-        .operation(OperationType.CREATE)
-        .resourcePath(session.getContext().getUri(), or.getName())
-        .representation(or)
-        .success();
+            .resource(ORGANIZATION_ROLE.name())
```
Comment (xgp) [inline src/main/java/io/phasetwo/service/resource/RolesResource.java:103]: Need a test that validates we can emit multiple events per request.

File: src/main/java/io/phasetwo/service/resource/UserResource.java [hunk 6]
```diff
@@ -59,6 +66,79 @@ public Stream<OrganizationRole> listUserOrgRoles(
...
+    return Response.created(session.getContext().getUri().getAbsolutePathBuilder().build())
+            .build();
+  }
+
+  @DELETE
+  @Path("/{userId}/orgs/{orgId}/roles")
+  @Consumes(MediaType.APPLICATION_JSON)
+  public Response revokeUserOrgRoles(@PathParam("userId") String userId, @PathParam("orgId") String orgId,
+                                     List<OrganizationRole> rolesRep) {
+    log.debugv("Revoke user organization roles for %s %s %s", realm.getName(), userId, orgId);
```
Comment (xgp) [inline src/main/java/io/phasetwo/service/resource/UserResource.java:103]: this should be `debugf`

File: src/main/java/io/phasetwo/service/resource/UserResource.java [hunk 7]
```diff
@@ -59,6 +66,79 @@ public Stream<OrganizationRole> listUserOrgRoles(
     }
   }
 
+  @PUT
+  @Path("/{userId}/orgs/{orgId}/roles")
+  @Consumes(MediaType.APPLICATION_JSON)
+  public Response grantUserOrgRoles(@PathParam("userId") String userId, @PathParam("orgId") String orgId,
+                                    List<OrganizationRole> rolesRep) {
+    log.debugv("Grant user organization roles for %s %s %s", realm.getName(), userId, orgId);
```
Comment (xgp) [inline src/main/java/io/phasetwo/service/resource/UserResource.java:74]: this should be `debugf`

File: src/main/java/io/phasetwo/service/resource/RolesResource.java [hunk 8]
```diff
@@ -41,12 +44,52 @@ public Stream<OrganizationRole> getRoles() {
...
+
+    representation.forEach(this::createOrganizationRole);
+
+    return Response.created(session.getContext().getUri().getAbsolutePathBuilder().build())
+            .build();
+  }
+
+  @DELETE
+  @Consumes(MediaType.APPLICATION_JSON)
+  public Response deleteRoles(List<OrganizationRole> representation) {
```
Comment (xgp) [inline src/main/java/io/phasetwo/service/resource/RolesResource.java:69]: Needs a test.

File: src/main/java/io/phasetwo/service/resource/RolesResource.java [hunk 9]
```diff
@@ -41,12 +44,52 @@ public Stream<OrganizationRole> getRoles() {
...
+  }
+
+  @PUT
+  @Consumes(MediaType.APPLICATION_JSON)
+  public Response createRoles(List<OrganizationRole> representation) {
+    canManage();
+
+    representation.forEach(this::createOrganizationRole);
+
+    return Response.created(session.getContext().getUri().getAbsolutePathBuilder().build())
```
Comment (xgp) [inline src/main/java/io/phasetwo/service/resource/RolesResource.java:63]: This doesn't really give the user information about what was created. In the single case, this returns a path that is unique to the newly created resource. What do other APIs do when there is a bulk method?

File: src/main/java/io/phasetwo/service/resource/UserResource.java [hunk 10]
```diff
@@ -59,6 +66,79 @@ public Stream<OrganizationRole> listUserOrgRoles(
     }
   }
 
+  @PUT
+  @Path("/{userId}/orgs/{orgId}/roles")
+  @Consumes(MediaType.APPLICATION_JSON)
+  public Response grantUserOrgRoles(@PathParam("userId") String userId, @PathParam("orgId") String orgId,
```
Comment (xgp) [inline src/main/java/io/phasetwo/service/resource/UserResource.java:72]: Needs a test.

File: src/main/java/io/phasetwo/service/resource/UserResource.java [hunk 11]
```diff
@@ -59,6 +66,79 @@ public Stream<OrganizationRole> listUserOrgRoles(
...
+                                    List<OrganizationRole> rolesRep) {
+    log.debugv("Grant user organization roles for %s %s %s", realm.getName(), userId, orgId);
+    UserModel user = session.users().getUserById(realm, userId);
+    OrganizationModel org = orgs.getOrganizationById(realm, orgId);
+    canManage(userId, orgId, user, org);
+    rolesRep.forEach(r -> {
```
Comment (xgp) [inline src/main/java/io/phasetwo/service/resource/UserResource.java:78]: Need a test that validates when some of the grants fail:
- everything fails
- no events get emitted
- a proper error gets sent back

File: src/main/java/io/phasetwo/service/resource/UserResource.java [hunk 12]
```diff
@@ -59,6 +66,79 @@ public Stream<OrganizationRole> listUserOrgRoles(
...
+
+        adminEvent
+                .resource(ORGANIZATION_ROLE_MAPPING.name())
+                .operation(OperationType.CREATE)
+                .resourcePath(session.getContext().getUri())
+                .representation(userId)
+                .success();
+      }
+    });
+    return Response.created(session.getContext().getUri().getAbsolutePathBuilder().build())
```
Comment (xgp) [inline src/main/java/io/phasetwo/service/resource/UserResource.java:94]: As above, this is not informative/unique. Need to figure out a better response.

File: src/main/java/io/phasetwo/service/resource/UserResource.java [hunk 13]
```diff
@@ -59,6 +66,79 @@ public Stream<OrganizationRole> listUserOrgRoles(
...
+            .build();
+  }
+
+  @DELETE
+  @Path("/{userId}/orgs/{orgId}/roles")
+  @Consumes(MediaType.APPLICATION_JSON)
+  public Response revokeUserOrgRoles(@PathParam("userId") String userId, @PathParam("orgId") String orgId,
```
Comment (xgp) [inline src/main/java/io/phasetwo/service/resource/UserResource.java:101]: Needs a test.

File: src/main/java/io/phasetwo/service/resource/UserResource.java [hunk 14]
```diff
@@ -59,6 +66,79 @@ public Stream<OrganizationRole> listUserOrgRoles(
...
+                                     List<OrganizationRole> rolesRep) {
+    log.debugv("Revoke user organization roles for %s %s %s", realm.getName(), userId, orgId);
+    UserModel user = session.users().getUserById(realm, userId);
+    OrganizationModel org = orgs.getOrganizationById(realm, orgId);
+    canManage(userId, orgId, user, org);
+    rolesRep.forEach(r -> {
```
Comment (xgp) [inline src/main/java/io/phasetwo/service/resource/UserResource.java:107]: As above, need a test that validates when some of the deletes fail:
- everything fails
- no events get emitted
- a proper error gets sent back

**Review by xgp — CHANGES_REQUESTED** [review]

**Comment by farafonov-alexey** [comment]
**Deleted /with-roles method.**

Still couldn't understand how to get this info in two API calls.
- we can search users with some criteria - result list of users (1st call)
- we can get list of org roles - result list of roles (2nd call)

But how we can match users to roles?

**Comment by xgp** [comment]
> Still couldn't understand how to get this info in two API calls.
> 
> * we can search users with some criteria - result list of users (1st call)
> * we can get list of org roles - result list of roles (2nd call)
> 
> But how we can match users to roles?

I'm not exactly sure what you want to do @farafonov-alexey 
1. This will get a paginated list of users who are a member of the specified organization. https://phasetwo.io/api/get-organization-memberships
2. This will check if a user is a member of an organization. https://phasetwo.io/api/check-organization-membership
3. This will list organization roles for the given user and org. https://phasetwo.io/api/list-organization-roles-for-the-given-user-and-org

Is there another case I am missing?

**Comment by farafonov-alexey** [comment]
> Is there another case I am missing?

We have a webpage which contains the table.

| Username  | email | ... other user info ... | User roles |
| ------------- | ------------- |------------- |------------- |
| User1  | user1@mail.com  | ... | Role5, Role6, Role7 |
| User2  | user2@mail.com  | .... | Role1, Role2 |
| ... | ....  | .... | .... |
| UserN  | userN@mail.com  | .... | Role4, Role8 |


We can easily get the list of users with the help of membership call - https://phasetwo.io/api/get-organization-memberships

But in order to get the last column roles info, we need to make N calls (https://phasetwo.io/api/list-organization-roles-for-the-given-user-and-org) for each specific user (User1, User2 ... UserN).

**Comment by farafonov-alexey** [comment]
Refactor response for bulk operations - https://tyk.io/blog/api-design-guidance-bulk-and-batch-import/
Now if we create 1 already existing role and 1 new role we get this kind of response.
```
207 Multi-Status

[ {
  "status" : 400,
  "error" : "HTTP 409 Conflict",
  "item" : {
    "name" : "eat-apples"
  }
}, {
  "status" : 201,
  "item" : {
    "name" : "drink-coffee"
  }
} ]
```
That's for discussion.

Also can't figure out how to test `no events get emitted`, maybe any example? @xgp

**Comment by farafonov-alexey** [comment]
Any updates about this? @xgp

**Comment by xgp** [comment]
Is it ready to be reviewed again?

**Comment by farafonov-alexey** [comment]
I'm currently waiting for your response on 2 questions - 
1. Is this bulk answer is ok? - https://github.com/p2-inc/keycloak-orgs/pull/99#issuecomment-1630811235. If yes I will rewrite other EPs like this
2. How to test `no events get emitted`? Any example?

**Comment by farafonov-alexey** [comment]
> 2\. there is an example of how to set up a test for event listeners (using http) here https://github.com/p2-inc/keycloak-events/blob/main/src/test/java/io/phasetwo/keycloak/resources/EventsResourceTest.java#L45

I'm trying to work with webhooks API right now here -
https://github.com/farafonov-alexey/keycloak-orgs/blob/b6ec2dea947ebcf5affac2ee0957282d49c326ea/src/test/java/io/phasetwo/service/resource/OrganizationResourceTest.java#L655

But  `https://phasetwo.io/api/get-webhooks` as well as `https://phasetwo.io/api/create-webhook` API calls constantly returning this response:
`
response webhooks: {
  "error" : "HTTP 404 Not Found"
}
`

Any clues what is wrong with this code.

**Comment by farafonov-alexey** [comment]
> 1. The events extension has to be installed
Ok, that makes sense. How to install this extensions in tests?

**Comment by xgp** [comment]
Should just be necessary to add it as a test dependency in the pom.xml.

**Comment by farafonov-alexey** [comment]
> @farafonov-alexey LMK when it's ready to review/test again.

Sure! Webhooks tests are working now and I get adminEvents as expected 👍 

One more thing. I'm testing delete methods with SimpleHttp here - https://github.com/farafonov-alexey/keycloak-orgs/blob/a9c0fcb7ead49e8e725892656a7a9338311e2e85/src/test/java/io/phasetwo/service/resource/OrganizationResourceTest.java#L706 and get this error
```
{
  "error" : "RESTEASY003065: Cannot consume content type"
}
```

If I build a jar and call this API from Postman everything is fine.

Maybe `SimpleHttp.doDelete.toJson` works differently?

**Comment by xgp** [comment]
I don't have a quick explanation. Have you tried using something like Wireshark to look at the raw HTTP that Postman and SimpleHttp are sending to see how they differ?

**Comment by farafonov-alexey** [comment]
Hello, @xgp! Any updates on this?

**Comment by farafonov-alexey** [comment]
Hello, @xgp!
Do you still interested in this PR?

**Comment by xgp** [comment]
Thanks for pinging and assigning this. We need to 
- resolve conflicts with the current state of `main`
- document the changes in the REAME (see the new "active organization" documentation for a good example)
- update the [openapi.yml](https://github.com/p2-inc/phasetwo-docs/blob/main/openapi.yaml) with the API changes

**Comment by farafonov-alexey** [comment]
Hello! @xgp
I'm currently in resolving conflicts stage.
But constantly getting test error's like below.
```
[ERROR] io.phasetwo.service.resource.OrganizationResourceTest.testAddGetDeleteRolesBulk -- Time elapsed: 0.091 s <<< FAILURE!
java.lang.AssertionError: 

Expected: is <201>
     but: was <404>
```

It seems that my changes doesn't affect test stage. Maybe I'm missing some _testcontainers_ functionality. What is the proper workflow to test new code?

File: src/main/java/io/phasetwo/service/resource/UserResource.java [hunk 15]
```diff
@@ -14,57 +16,62 @@
...
+                return org.getRolesStream()
+                        .filter(r -> r.hasRole(user))
+                        .map(r -> convertOrganizationRole(r));
+            } else {
+                throw new NotFoundException("User is not a member of the organization");
+            }
+        } else {
+            throw new NotAuthorizedException("Insufficient permissions");
+        }
     }
```
Comment (xgp) [inline src/main/java/io/phasetwo/service/resource/UserResource.java:74]: This seems like just a formatting change. Can you revert?

**Review by xgp — CHANGES_REQUESTED** [review]
Everything looks good except the formatting-only changes in `UserResource`. Please fix that and I'll merge. Thank you for your work and patience!

File: src/main/java/io/phasetwo/service/resource/UserResource.java (hunk 15 above)
Comment (farafonov-alexey) [reply src/main/java/io/phasetwo/service/resource/UserResource.java]: Done =)

**Review by xgp — APPROVED** [review]

**Comment by farafonov-alexey** [comment]
Found a copy-paste in bulk-roles.md. Fixed.

**Review by xgp — APPROVED** [review]

**Pull request merged**